File: 400-y.rb
fact = ->(f) {
  ->(x) {
    if x == 1
      1
    else
      x * f.(f).(x-1)
    end
  }
}.(
  ->(f) {
    ->(x) {
      if x == 1
        1
      else
        x * f.(f).(x-1)
      end
    }
  }
)
puts fact.(8)

#Refactor - there's some duplication, extract the re-application of the function to itself

File: js/400-y.js
const fact = (f) => {
  return (x) => {
    if (x === 0) {
      return 1
    } else {
      return x * f(f)(x-1)
    }
  }
}((f) => {
    return (x) => {
      if (x === 0) {
        return 1
      } else {
        return x * f(f)(x-1)
      }
    }
  }
)

console.log(fact(8))

//Refactor - there's some duplication, extract the re-application of the function to itself
